compiler/src/iree/compiler/Codegen/Common/test/math_transform.mlir

```diff
@@ -55,3 +55,39 @@ func.func @rewrite_erf(%arg0: f16) -> f16 attributes {
   %0 = math.erf %arg0 : f16
   return %0 : f16
 }
+
+// -----
+
+// CHECK-LABEL: @no_approx_on_rocm
+func.func @no_approx_on_rocm(%arg0: f16) -> f16 attributes {
+  hal.executable.target =  #hal.executable.target<"rocm", "rocm-hsaco-fb", {}>
+} {
+  // On ROCm, we want to use the native device library functions.
+  // It's OK for f16 to still get casted to f32, as
+  // the device library functions for f16 are casting to f32 anyway.
+  // CHECK:         math.acos
+  // CHECK:         math.atan
+  // CHECK:         math.sin
+  // CHECK:         math.tanh
+  // CHECK:         math.log
+  // CHECK:         math.log2
+  // CHECK:         math.log1p
+  // CHECK:         math.exp
+  // CHECK:         math.exp2
+  // CHECK:         math.expm1
+  // CHECK:         math.cbrt
+  // CHECK:         math.erf
```

Comment:
have we figured out the numerical issue with `math.erf` library function?

Reply:
We have (99%) figured that there was no issue with it, and the issues we ran into were caused by PolynomialApproximationPass being too coarse-grained and too convoluted, so that when we thought earlier that we were enabling/disabling math.erf approximation, we were also enabling/disabling a number of other things, unintentionally. This is what https://github.com/iree-org/iree/pull/19922 solved. Now in the present PR we are finally at a place where we have some fine-grained, well-defined levers to play with.

Reply:
The remaining issue has been diagnosed in https://github.com/iree-org/iree/pull/20074#issuecomment-2679531327 as an issue with overly strict tests requiring agreement with the less accurate polynomial approximations in f16, as opposed to the ROCm device lib performing the approximation after upcasting to f32.